### ISSUE TITLE
Program never stops

### ISSUE THREAD

**Issue by chekanov** [AIDASoft/DD4hep#1027]
Hi,
 
 I'm not sure is this is a bug or not, but I'm posting this problem here. I'm trying to design a simple ECAL with PbWO4 crystals, using the CLIC calorimeter geometry. If I just use PbWO4 without optical properties, everything works. When optical properties are added to PbWO4 , the program runs forever. 

Here is GIT to reproduce this problem: https://github.com/chekanov/SimpleECAL  

Note that this example is similar  to SingleDualCrystal (used by Sarah Eno), which works fine for single crystals. 

This infinite loop problem can be fixed by commenting out 

` scint = PhysicsList(kernel, 'Geant4ScintillationPhysics/ScintillationPhys')`

 but keeping `Geant4CerenkovPhysics/CerenkovPhys` and `Geant4OpticalPhotonPhysics/OpticalGammaPhys` in the steering card  `SCEPCALsteering.py`

 - OS version: e.g. centos7
 - Compiler version: e.g GCC 11.0
 - Package version: tag or commit ID, or GitHub branch
 - Reproduced by:  See the GIT https://github.com/chekanov/SimpleECAL  t
 - Goal: Making ECAL geometry with crystals

**Comment by MarkusFrankATcernch** [comment]
I have built the example. How do I start the simulation ?

**Comment by chekanov** [comment]
You can simply run a script from "compact":

`> A_RUN`

that makes a ROOT file or interactive method:

`> A_INTERACTIVE`

Then run:

```
/control/execute vis.mac
/run/beamOn 1
```

**Comment by MarkusFrankATcernch** [comment]
Is the `A_INTERACTIVE` in the repository supposed to reproduce the problem?
With the current version I was happily running:
```
DRcalo deposit 0 position (43,52,0) string /ECalBarrel_0/stave_6/layer4_3/slice3_2 volume id 14636883489332340
DRcalo deposit 0 position (28,47,0) string /ECalBarrel_0/stave_6/layer4_3/slice2_1 volume id 13229444181269588
DRcalo deposit 0 position (28,48,0) string /ECalBarrel_0/stave_6/layer4_3/slice3_2 volume id 13510919157980276
DRcalo deposit 0 position (26,44,0) string /ECalBarrel_0/stave_6/layer4_3/slice2_1 volume id 12385010661203028
DRcalo deposit 0.069689 position (25,42,0) string /ECalBarrel_0/stave_6/layer4_3/slice2_1 volume id 11822056412814420
DRcalo deposit 1.00183 position (25,41,0) string /ECalBarrel_0/stave_6/layer4_3/slice2_1 volume id 11540581436103764
DRcalo deposit 1.91461 position (25,43,0) string /ECalBarrel_0/stave_6/layer4_3/slice2_1 volume id 12103531389525076
DRcalo deposit 0 position (28,45,0) string /ECalBarrel_0/stave_6/layer4_3/slice2_1 volume id 12666494227848276
DRcalo deposit 1.71739 position (29,44,0) string /ECalBarrel_0/stave_6/layer4_3/slice2_1 volume id 12385023546104916
DRcalo deposit 0 position (29,45,0) string /ECalBarrel_0/stave_6/layer4_3/slice2_1 volume id 12666498522815572
DRcalo deposit 0 position (30,45,0) string /ECalBarrel_0/stave_6/layer4_3/slice2_1 volume id 12666502817782868
GenerationInit   WARN  +++ Finished run 1 after 5 events (6 events in total)
0 events have been kept for refreshing and/or reviewing.
  "/vis/reviewKeptEvents" to review them one by one.
  "/vis/enable", then "/vis/viewer/flush" or "/vis/viewer/rebuild" to see them accumulated.
Idle> exit
Graphics systems deleted.
Visualization Manager deleting...

Geant4Kernel     INFO  ++ Terminate Geant4 and delete associated actions.
RootOutput       INFO  +++ Closing ROOT ourpur file testSCEPCAL.root
DDSim            INFO DDSim            INFO  Total Time:   18.87 s (User), 1.10 s (System)
[Thread 0x7fd9e78e2700 (LWP 228062) exited]
[Inferior 1 (process 228004) exited normally]
(gdb) q
```
... and thanks for letting me spot the type: `Closing ROOT ourpur file`....
If such typos are spotted everyone should somehow bring it up.

**Comment by chekanov** [comment]
Yes, I posted a working version with  PbWO4 defined without the optical properties.  Please change PbWO4 to E_PbWO4 on the line 42 of compact/ECalBarrel_DualCrystal.xml and run this example again.  Then this iterative example will run forever. I suggest to have a second terminal to kill the program otherwise you may need a hard reboot. Note that we tested E_PbWO4 using a single crystal geometry and such an example works. The problem appears when we make a full ECAL with the crystals organized as PolyhedraRegular.

**Comment by MarkusFrankATcernch** [comment]
Do you then get output like the following forever:
```
DRcalo deposit 2.60848e-06 position (13,168,0) string /ECalBarrel_0/stave_8/layer1_0/slice2_1 volume id 47287851926168660
DRcalo deposit 2.60848e-06 position (40,114,0) string /ECalBarrel_0/stave_8/layer1_0/slice2_1 volume id 32088319147910228
DRcalo deposit 2.60848e-06 position (34,102,0) string /ECalBarrel_0/stave_8/layer1_0/slice1_0 volume id 28710593657578548
DRcalo deposit 2.72093e-06 position (-10,-55,0) string /ECalBarrel_0/stave_8/layer1_0/slice2_1 volume id -15199691687844780
DRcalo deposit 2.63858e-06 position (35,138,0) string /ECalBarrel_0/stave_8/layer1_0/slice2_1 volume id 38843697114129492
DRcalo deposit 2.63858e-06 position (14,175,0) string /ECalBarrel_0/stave_8/layer1_0/slice3_2 volume id 49258181058110580
DRcalo deposit 2.65e-06 position (-39,51,0) string /ECalBarrel_0/stave_8/layer1_0/slice2_1 volume id 14636531289433172
DRcalo deposit 2.71359e-06 position (0,-36,0) string /ECalBarrel_0/stave_7/layer1_0/slice2_1 volume id -10133099157381036
DRcalo deposit 2.71359e-06 position (41,-49,0) string /ECalBarrel_0/stave_8/layer2_1/slice1_0 volume id -13792097756765132
DRcalo deposit 2.65877e-06 position (37,-3,0) string /ECalBarrel_0/stave_9/layer1_0/slice3_2 volume id -844266012137356
DRcalo deposit 2.62378e-06 position (-2,105,0) string /ECalBarrel_0/stave_8/layer1_0/slice2_1 volume id 29836338945598548
DRcalo deposit 2.83043e-06 position (5,-103,0) string /ECalBarrel_0/stave_8/layer1_0/slice3_2 volume id -28991901122157452
DRcalo deposit 2.44291e-06 position (26,61,0) string /ECalBarrel_0/stave_8/layer1_0/slice2_1 volume id 17170085252703316
DRcalo deposit 2.44291e-06 position (17,127,0) string /ECalBarrel_0/stave_8/layer1_0/slice1_0 volume id 35747395060900916
DRcalo deposit 2.72603e-06 position (1,-32,0) string /ECalBarrel_0/stave_8/layer1_0/slice1_0 volume id -9007194955570124
^C
Thread 1 "python3.9" received signal SIGINT, Interrupt.
0x00007f79aa723dc1 in G4String::operator== (this=0xd10b8a0, str=...)
    at /cvmfs/sft.cern.ch/lcg/views/LCG_101/x86_64-centos7-gcc11-dbg/include/Geant4/G4String.icc:119
119	}
```
For me this kind of output continues forever....
Is this what you mean ?

**Comment by chekanov** [comment]
Yes, exactly. It prints forever without stop.

**Comment by MarkusFrankATcernch** [comment]
Hmmm. This does not look immediately like a DD4hep problem, since the Geant4 kernel keeps on tracking without ever stopping. If I remove in the material description:
```
  <material name="E_PbWO4">
    <D value="8.28" unit="g/cm3" />
    <fraction n="0.45532661" ref="Pb"/>
    <fraction n="0.40403397" ref="W"/>
    <fraction n="0.14063942" ref="O"/>
    <property name="RINDEX"        ref="RINDEX__PbWO4"/>
      <property name="ABSLENGTH" ref="AbsLen_PS"/>
<!--
      <property name="FASTCOMPONENT" ref="scintFast_PS"/>
-->
      <constant name="SCINTILLATIONYIELD" value="13.9/keV"/>
      <constant name="FASTTIMECONSTANT" value="2.8*ns"/>
      <constant name="RESOLUTIONSCALE" value="1."/>
  </material>
```
the FASTCOMPONENT, then the simulation terminates, but creates many more and much smaller deposits 
than without any property. 
I think what is needed here is a better understanding what these properties really mean for Geant4.
DD4hep is here only the mediator. 
I checked and according to my understanding are values are passed correctly to Geant4.

**Comment by chekanov** [comment]
It it possible the the light is reflected on the wall of   PolyhedraRegular geometry and travels around in the phi? Maybe there is a way to insulate sides of the crystals ? I'm not sure how this can be done since I can only see how to insulate  4 crystals in the depth, but not on the sides..  A similar example for a  module made from several such crystals works fine (example by Sarah Eno https://github.com/saraheno/DualTestBeam). This problem happens when we arrange the crystals using PolyhedraRegular , i.e.    ECAL barrel,  circular in Phi,

**Comment by MarkusFrankATcernch** [comment]
Hmm. Then the problem should disappear if only one stave would be constructed. This would eliminate this bit.

**Comment by MarkusFrankATcernch** [comment]
You will have to add a line
if ( i == <X> ) break;
at line 156 of EcalBarrel_geo.cpp, where <X> is the number of staves you actually want.

changing "nsides" in xml alone will not do the job.

**Comment by saraheno** [comment]
So Sergei: my attempt to insulate the sides of the crystals is being discussed in this thread: #1014 
dd4hep does have the possibility of putting optical surfaces between optical elements, but somehow it is not quite working.  I am disucssing with Andre and with Hans Wenzel if this is a problem with dd4hep or geant4.

**Comment by hanswenzel** [comment]
Hi Sarah et al.


the provided gdml file is not valid as far as Geant4 (I was using the latest version of Geant4)  is concerned. Reading it in from CaTS I get:

------- EEEE ------- G4Exception-START -------- EEEE -------
*** G4Exception : ReadError
      issued by : G4GDMLReadDefine::getMatrix()
Matrix 'SCINTILLATIONYIELD_0x2dfbf80' was not found!
*** Fatal Exception *** core dump ***
 **** Track information is not available at this moment
 **** Step information is not available at this moment

-------- EEEE -------- G4Exception-END --------- EEEE -------

that makes sense since SCINTILLATIONYIELD is defined as a constant
   <constant name="SCINTILLATIONYIELD_0x2dfbf80" value="13899992.686390091"/>

while the property requires a matrix:

      <property name="SCINTILLATIONYIELD" ref="SCINTILLATIONYIELD_0x2dfbf80"/>


What is the Geant4 version used by dd4hep?




Hans


--

Dr. Hans Wenzel



Physics and Detector Simulation

Fermi National Accelerator Laboratory

P.O. Box 500, MS 223

Batavia, Illinois 60510

USA



630 840 6034 office

331 250 9342  mobile

home.fnal.gov/~wenzel<http://home.fnal.gov/~wenzel>

***@***.******@***.***>

________________________________

**Comment by hanswenzel** [comment]
Hi Sarah,


that's definitely recent enough. So there is definitely something wrong in the compact to gdml conversion. It's not only the SCINTILLATIONYIELD but other properties like e.g. the FASTTIMECONSTANT that should be defined as matrices not constants.

 Another problem in the gdml file is:

    <bordersurface name="/world/DRCrystal#HallCrys" surfaceproperty="/world/DRCrystal#mirrorSurface">
      <physvolref ref="tower_0" ref="DRCrystal_0"/>
    </bordersurface>

a bordersurface needs 2 physical  volumes (compared to a skin surface that only needs 1 logical volume) e.g:

<bordersurface name="bordersrf0" surfaceproperty="surf1" >
<physvolref ref="pCalorimeterVolume"/>
<physvolref ref="pPhotonVolume0"/>
</bordersurface>


So I don't know dd4hep but the gdml file that is produced and should represent the Geant4 geometry definitely is not valid.
I modified the gdml file and attached it to this mail as well as a screenshot of CaTS reading in this file. I will shoot some optical photons into this geometry to see if the surface works properly but I think Geant4 is just fine.
CaTS might be a better solution for single crystals or test beam set ups 🙂 .

Hans


--

Dr. Hans Wenzel



Physics and Detector Simulation

Fermi National Accelerator Laboratory

P.O. Box 500, MS 223

Batavia, Illinois 60510

USA



630 840 6034 office

331 250 9342  mobile

home.fnal.gov/~wenzel<http://home.fnal.gov/~wenzel>

***@***.******@***.***>

________________________________

**Comment by saraheno** [comment]
@hanswenzel yes, we have a team at Virginia working on CaTS.  But eventually we need to build a full detector that will work with the CLIC/ILC/CEPC software, so it has to be in dd4hep.  We cannot build an entire detector from scratch.  We want to use other's trackers etc.  So I'm doing the single bars etc to understand better the code, working with Sergei to eventually get a full detector in dd4hep. while the Virginia team uses CaTS for our cosmic ray tests that are starting now. (anyway, right now dd4hep is too slow...  I hope somehow we can get your accelerator code into it or we may only be able to put in parameterizations of the optics, not use the GEANT4 optics)

@MarkusFrankATcernch @andresailer The place where I attach the surface is at https://github.com/saraheno/SingleDualCrystal/blob/6589de4b77b686399b7236cf3d33de8f708833f1/src/DRCrystal_geo.cpp#L213

i think I am attaching it between pv and env_phv.  is that not right?

**Comment by MarkusFrankATcernch** [comment]
@saraheno If the gdml is not correct: how did you produce it?
- using ROOT TGeo (with the dd4hep plugin)
- or Geant4 using `/ddg4/ConstructGeometry/writeGDML  <file-name>`  from the Geant4 prompt ?

The surface stuff in ROOT is rather recent (and the surface gdml writing probably not extensively debugged), 
but the command from the Geant4 prompt invokes the Geant4 GDML writer.
This could be an explanation which the GDML is inconsistent.

**Comment by hanswenzel** [comment]
Hi I just checked by dumping the geometry that I constructed with my valid gdml file directly from geant4 memory to a new gdml file and the produced gdml file just looks fine.  The surfaces are defined as skinsurfaces and the relevant properties are defined as matrices. So I don't think the problem is Geant4  


Hans

**Comment by saraheno** [comment]
@hanswenzel when you run muons through this, making cherenkov photons, do any escape through the surface?  or are they all trapped?

**Comment by hanswenzel** [comment]
Hi Sarah 

Haven't tried that yet . I will try and let you know after lunch.  I just shot single optical photons. But after looking at the gdml files one first needs  to actually fix the detector description.  I have put 2 files into github: 
https://github.com/hanswenzel/CaTS/tree/main/gdml

one is DRSingleCrystal.gdml the other is sarah.gdml. The first one is the corrected gdml file that Sarah sent me the second  is the dump of the geometry from Geant4 into a gdml file. Both are valid and can be used by Geant4 as input.

**Comment by saraheno** [comment]
@hanswenzel I understood part of the problem was that the surface in the gdml was only attached to one volume.  to fix it, you must have attached it to a second volume? to which volume did you attach it?

**Comment by hanswenzel** [comment]
Hi there were actually several problems with the geometry: 1) constants are used when matrices are required, 2) There are 2 types of surfaces a border surface between two physical volumes (e.g. between the crystal and the photodetector) and a skin surface which defines the surface (all sides) of a "logical" volume. The border surface requires 2 physical volumes as input the skin surface one logical volume. See the gdml files that I put into the CaTS github.  You will need both types (crystal to surrounding  air (skin)  and  the end of the crystal  surface needs to be interfaced with a photo detector surface(border) . It looks to me that the dd4hep implementation needs some work. The produced gdml file is not a valid gdml file. The 2 gdml files in github show the input gdml and the direct dump from Geant4 memory. I had problems in the past were the root gdml parser was slightly different from the Geant4 one creating all kinds of problems.

**Comment by saraheno** [comment]
@hanswenzel in your previous post, could you defined better "input gdml" and "direct dump from GEANT4 memory"? I'm not sure exactly what they are.

**Comment by hanswenzel** [comment]
Hi Sarah, 
 the file 
https://github.com/hanswenzel/CaTS/blob/main/gdml/DRSingleCrystal.gdml
is basically the file you provided and that I modified to actually make it a valid gdml file. 
The file: https://github.com/hanswenzel/CaTS/blob/main/gdml/sarah.gdml
is the dump of the Geant4 memory to gdml which is equivalent to the input. I just wanted to make sure that nothing got corrupted between reading and writing and the Geant4 parser and writer are fine. On the other hand the original file produced by the geomconverter didn't produce a valid gdml file and the output is not a representation of a valid Geant4 geometry


Hans
.

**Comment by saraheno** [comment]
Hi Hans,
I think the dd4hep team is still waiting for you to reply to
https://github.com/AIDASoft/DD4hep/issues/1027#issuecomment-1339810988

**Comment by MarkusFrankATcernch** [comment]
@saraheno 
- If you issue `/ddg4/ConstructGeometry/writeGDML <file-name>` the GDML is produced from the Geant4 memory.
- Using `geoConverter -compact2gdml -input DRSingleCrystal.xml -output DRSingleCrystal.gdml` invokes the ROOT GDML writer.

I need to know if `/ddg4/ConstructGeometry/writeGDML <file-name>` produces a correct GDML to know if the problem is in the conversion to Geant4 or in the ROOT GDML writer. This would help to fix the problem.

**Comment by saraheno** [comment]
my apologies @MarkusFrankATcernch : I do not seem to understand properly how to do that command

![image](https://user-images.githubusercontent.com/5215288/206186795-0ff6a929-1504-47b7-b06b-22259627a698.png)

**Comment by andresailer** [comment]
Hi @saraheno 
Run ddsim with `--runtype shell` and put the command in the geant4 shell

**Comment by saraheno** [comment]
When I did that, I got 

![image](https://user-images.githubusercontent.com/5215288/206189199-4d1f7c7a-d8f2-407b-af68-ffa36e19fd48.png)

the command I gave was
 ddsim --runtype shell

**Comment by andresailer** [comment]
`--runType` with capital T, sorry!

**Comment by andresailer** [comment]
You also need to give all the other arguments to build the detector, e.g., `--compactFile`

**Comment by saraheno** [comment]
do I did:
 ddsim --steeringFile SCEPCALsteering.py --compact ./DRSingleCrystal.xml --runType shell --part.userParticleHandler='' -G --gun.position="0.,30.,0." --gun.direction "0 -1 0" --gun.energy "1*GeV" --gun.particle="mu-" -N 1 -O out.root

and got 

![image](https://user-images.githubusercontent.com/5215288/206190667-961d8796-749f-4cf0-b5cf-6b76a52f28a5.png)

**Comment by saraheno** [comment]
nevermind
I'll try again

**Comment by saraheno** [comment]
okay, it worked.  The gdml file created that way is at: https://github.com/saraheno/SingleDualCrystal/blob/main/compact/DRSingleCrystal.gdml

**Comment by MarkusFrankATcernch** [comment]
@saraheno Thank you. I will check the output GDML. 
If things are correct, it should be the same as the corrected GDML from Hans.
Do you have somewhere Hans' GDML in git ?

**Comment by saraheno** [comment]
@hanswenzel @MarkusFrankATcernch 

I think the file Hans made by correcting my old file is at https://github.com/hanswenzel/CaTS/blob/main/gdml/DRSingleCrystal.gdml

or at

https://github.com/hanswenzel/CaTS/blob/main/gdml/sarah.gdml

**Comment by saraheno** [comment]
@hanswenzel when I look at this display, it looks like photons are escaping out the sides even though it is wrapped with a perfectly reflective surface?  Is this what is happening?

**Comment by MarkusFrankATcernch** [comment]
I compared the 2 GDMLs:
- the ROOT GDML output for surfaces needs fixing. Please do not use this converter for the time being.
- Dumping from Geant4 I get:
```
    <opticalsurface finish="0" model="0" name="_world_DRCrystal_mirrorSurface0xeb35ed0" type="1" value="1">
      <property name="REFLECTIVITY" ref="REFLECTIVITY0xeb3c0b0"/>
      <property name="EFFICIENCY" ref="EFFICIENCY0xeb44290"/>
    </opticalsurface>
....
    <bordersurface name="_world_DRCrystal_HallCrys0xeb66b40" surfaceproperty="_world_DRCrystal_mirrorSurface0xeb35ed0">
      <physvolref ref="tower_00xeb66800"/>
      <physvolref ref="DRCrystal_00xeb668d0"/>
    </bordersurface>
```
which looks correct to me, since you are adding a BorderSurface.
However, the output is different to the changes of Hans, but I guess he interpreted it differently as a skinsurface:
```
     <opticalsurface name="/world/DRCrystal#mirrorSurface" model="glisur" finish="polished" type="dielectric_dielectric" value="1">
      <property name="REFLECTIVITY" ref="REFLECTIVITY__0x2d946b0"/>
      <property name="EFFICIENCY" ref="EFFICIENCY__0x2d944d0"/>
    </opticalsurface>
....
   <skinsurface name="/world/DRCrystal#HallCrys" surfaceproperty="/world/DRCrystal#mirrorSurface">
      <volumeref  ref="DRCrystal"/>
    </skinsurface>
 ```
Here I have a question for @hanswenzel if he by chance knows: Does GDML support the literals as attributes for e.g. `<opticalsurface model="..."...>` ?

**Comment by hanswenzel** [comment]
I think that's true .   Just for fun I replaced the optical surface with one off a tyvek wrapped one. You will see the effect is much different. So what I can do is looking up the best surfaces to use. I guess the slices in the end are supposed to represent photo detectors so it doesn't make sense that they also radiate. 

 <opticalsurface name="_world_DRCrystal_mirrorSurface0xcf11350" model="LUT" finish="etchedtyvekair" type="dielectric_LUT" value="1.0"/>
    
![Screenshot at 2022-12-07 08-37-29](https://user-images.githubusercontent.com/7925307/206207916-0b18e703-9d54-4a66-9736-1644b0e8b24b.png)

**Comment by chekanov** [comment]
I've just made one small check by adding Al foil between 4 crystals (in depth): https://github.com/chekanov/SimpleECAL/blob/main/compact/ECalBarrel_DualCrystal.xml

```
            <layer repeat="4" vis="ECalLayerVis">
               <slice material = "Al"            thickness = "0.94*mm" vis="InvisibleNoDaughters"/>
               <slice material = "killMedia" thickness = "0.2" sensitive="yes" limits="cal_limits" vis="CrystalEcalSensitiveVis"/>
               <slice material = "E_PbWO4" thickness = "50.0*mm" sensitive="yes" limits="cal_limits" vis="CrystalEcalSensitiveVis" radiator="yes" />
               <slice material = "killMedia" thickness = "0.2" sensitive="yes" limits="cal_limits" vis="CrystalEcalSensitiveVis"/>
          </layer>
```

The program still goes forever.  Could be an insulation from each site of the crystals will help.

**Comment by MarkusFrankATcernch** [comment]
13.9/keV = 13.9/1e-6 = 13900000 in TGeo (Energies in GeV)
and becomes  when translated to Geant4 (Energies in MeV): 
13.9/TGeo::keV * (TGeo::keV / G4::keV) = 13.9/G4::keV = 13.9 / 1e-3 = 13900

This is quite different from 0.00001.

**Comment by saraheno** [comment]
My apologies Markus: I am missing your point?  I have been telling Sergei that I don't think the code is in an infinite loop, that one just cannot simulate that many photons in that many crystals in a reasonable amount of time (without way the CaTs accelerator).  I told him we could prove that by setting the number of photons made very low and then the program would finish an event.  

so I think there is nothing really wrong with the code per say.  It is just not reasonable without accelerators to put in the real number of photons.  Either give up on the scintillation simulation of the crystals, or dial it down to something that can finish in a reasonable amount of time.  I think giving up on simulating the scintillation is best.  instead, use the energy deposits and then use a single crystal simulation to parameterize the number of collected photons as a function of the energy deposit size and position.  That is what CMS does for example.

**Comment by MarkusFrankATcernch** [comment]
@saraheno I am not sure about the program timing. I think there is nothing wrong with the code as well, but
there is simply a lack of understanding what the various parameters really mean and what their value should 
be when they enter Geant4 -- for some I even had to guess the units for the conversion from TGeo to Geant4.
If the parameters entering the game are far from what G4 expects, anything could go wrong.

For rechits or user defined hits, see `examples/DDG4_MySensDet` Have a look how the dictionary for the `MyTrackerhits`
is build. This should be all you need to e.g. save a `vector<MyTrackerHits(*)>` to a root tree.

**Comment by MarkusFrankATcernch** [comment]
There is more stuff coming up: 
I discovered that in Geant4 11 apparently FASTCOMPONENT and SLOWCOMPONENT are no longer material properties.
In Geant4 10.07 they were still there...

**Comment by hanswenzel** [comment]
That is correct one has to be aware that names of some optical properties changed between Geant4 10 and Geant4 11. You can still write gdml files that would work with both versions (the keywords for the different versions would just be ignored). Look e.g. at the https://github.com/hanswenzel/CaTS/blob/main/gdml/simpleLArTPC.gdml for a gdml file that would work both with 10 and 11. And I agree it's a pain in the bud.

**Comment by saraheno** [comment]
@MarkusFrankATcernch @hanswenzel 

Hans mentioned that one can set visual attributes to optical surfaces.  Is this possible to do in dd4hep?  I don't see this in the OpNovice example?

**Comment by hanswenzel** [comment]
The example how to extend the gdml schema and how to at visualization attributes to volumes can be found at:
https://geant4.kek.jp/lxr/source/examples/extended/persistency/gdml/G03/
In fact CaTS just uses this example.

**Comment by MarkusFrankATcernch** [comment]
@saraheno No. As far as I know neither the Geant4 nor the TGeo GDML writers support colors. I will have a look, but this is not the most urgent issue I suppose.

**Comment by MarkusFrankATcernch** [comment]
@saraheno  @chekanov  Can we close this issue ?
If I recapitulate correctly the problem was in correctly setting up Geant4 rather than in DD4hep per see.

**Comment by saraheno** [comment]
yes, that's right.  Please close it.
